website/docs/r/service_endpoint_policy_definition.html.markdown

```diff
@@ -0,0 +1,72 @@
+---
+subcategory: "Network"
+layout: "azurerm"
+page_title: "Azure Resource Manager: azurerm_service_endpoint_policy_definition"
+description: |-
+  Manages a Service Endpoint Policy Definition.
+---
+
+# azurerm_service_endpoint_policy_definition
+
+Manages a Service Endpoint Policy Definition.
```

Comment:
@katbyte Here is the [document](https://docs.microsoft.com/en-us/azure/virtual-network/virtual-network-service-endpoint-policies-overview) of this resource (and the policy one).

In short, from the document and the Swagger example, the policy definition currently only supports service `Microsoft.Storage`. It *allow you to filter egress virtual network traffic to Azure Storage accounts over service endpoint, and allow data exfiltration to only specific Azure Storage accounts*.

The `service_endpoint_name` here is always `Microsoft.Storage`.

The `service_resources` defines the allow list of storage accounts, which can be subscription scopes (e.g. /subscriptions/subscriptionId), or resource group scopes (e.g. /subscriptions/subscriptionId/resourceGroups/resourceGroupName) or individual accounts (e.g. /subscriptions/subscriptionId/resourceGroups/resourceGroupName/providers/Microsoft.Storage/storageAccounts/storageAccountName)

I'm following the same naming as used in Swagger here, and I guess it might be extended to support services other than Storage some day, that's why they picked some general naming here.

Reply:
hmm, maybe it makes sense to create a separate resource for each endpoint type supported? this one could be `azurerm_service_endpoint_storage_policy` and then the other resource `azurerm_service_endpoint_polices`?

also if these are always limited to subnets, would it make sense to put subnet in the resource name? ie `azurerm_subnet_service_endpoint_storage_policy`

Reply:
For the subnet part to apear in the name, I feel it is redundent since given the context of Azure, when you mention service endpoint, it is always referring to the concept of vnet.

For the separation, I don't see too much benefit to do it as it essentially just move the `service_endpoint_name` out to be part of the resource name. Additionally, the name `service endpoint policy` is also the advertise name chosen in Portal (see [this](https://docs.microsoft.com/en-us/azure/virtual-network/virtual-network-service-endpoint-policies-portal)).

So I prefer current interface..

Reply:
I think it would help for people new to azure and also group it alphabetically with subnets so there is an implied "this is what this applied to" but i'll leave that upto you. - my main point was making this specific to the storage, ie `azurerm_service_endpoint_storage_policy`

Reply:
If we move the `service_endpoint_name` out to be part of the resource name, then in the future if there is more service endpoint name allowed, then we will have to write more resource/data source and the test and document together. I think this burden makes the whole change less beneficial.

Reply:
I think the confusion here is that the naming is not necessarily in keeping with other policy approaches in Azure, where "definitions" are re-usable entities in their own right, rather than items belonging directly to a policy, and that the current resource/service support is limited to just the Storage RP.  It is closer to NSG in design approach.

If the Network team intends to follow this pattern in future for other Service Endpoints then this approach possibly just needs some consideration around property naming to help users understand what is meant by each property, and consider combining into one resource rather than 2, since the policy CreateOrUpdate supports inlining the "policy definitions" - WDYT? (It's probably worth you reaching out to the Network team to find out the intent here?)

Reply:
Just heard back from srevice team, turns out that they have no plan to expand Service Endpoint Policy to any other Service besides Storage. So I guess I'll do following changes per your suggestions:

- rename `azurerm_service_endpoint_policy` to `azurerm_subnet_service_endpoint_policy`
- merge the two resources into one
- remove the `service_endpoint_name` from the schema, but hardcode it to be Storage in the request